src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+				
+			/* 
+			 * Construct hash table used for shardId->Connection mapping.
+			 * We can not use connection cache from connection.c used by GteConnection because
+			 * we need to establish multiple connections with each nodes: one connection per shard
+			 */
+			shard2conn = CreateShard2ConnectionHash();
+
+			/* init state to read from COPY data source */
+			copyState = BeginCopyFrom(rel, copyStatement->filename,
+									  copyStatement->is_program,
+									  copyStatement->attlist,
+									  copyStatement->options);
+				
+			/* TODO: handle binary mode */
+			if (copyState->binary) { 
+				elog(ERROR, "Copy in binary mode is not currently supported");
+			}
+
+			PG_TRY();
+			{
+				while (nextRowFound)
+				{
+					MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+					nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+					MemoryContextSwitchTo(oldContext);
+					
+					/* write the row to the shard */
+					if (nextRowFound)
+					{		
+						if (columnNulls[partitionColumn->varattno]) { 
+							ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+											errmsg("cannot copy row with NULL value "
+												   "in partition column")));
+						} else { 
+							Datum partitionValue = columnValues[partitionColumn->varattno];
+							rightConst->constvalue = partitionValue;
+							prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+							
+							foreach(shardIntervalCell, prunedList)
+							{
+								ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+								int64 shardId = shardInterval->id;
+								bool found;
+								
+								copyConn = (CopyConnection*)hash_search(shard2conn, &shardInterval->id, HASH_ENTER, &found);
+								if (!found) { 
+									List *finalizedPlacementList;
+									copyConn->nReplicas = 0;
+									copyConn->preparedMask = 0;
+									/* grab shared metadata lock to stop concurrent placement additions */
+									LockShardDistributionMetadata(shardId, ShareLock);
+									
+									/* now safe to populate placement list */
+									finalizedPlacementList = LoadFinalizedShardPlacementList(shardId);
+									foreach(taskPlacementCell, finalizedPlacementList)
+									{
+										ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+										char *nodeName = taskPlacement->nodeName;
+										char nodePort[16];
+										PGconn* conn;
+										sprintf(nodePort, "%d", taskPlacement->nodePort);	 
+										conn = ConnectToNode(nodeName, nodePort);
+										if (conn != NULL)										 
+										{
+											copyConn->conn[copyConn->nReplicas++] = conn;
+											/*
+											 * New connection: start transaction with copy command on it.
+											 * Append shard id to table name.
+											 */
+											if (!tmgr->Begin(conn) ||
+												!PgShardExecute(conn, PGRES_COPY_IN, "%.*s_%ld%s", relationSuffixPos, query,
```

Comment:
What if I want to COPY from a file? The user would type FROM 'filename', but we should still send FROM STDIN to the worker.